File: guessing-game-bdd-atomic/src/test/java/tw/waterballsa/bddworkshop/stepdefs/then/Then_There_Is_N_Greeting_Records_In_The_Database.java
package tw.waterballsa.bddworkshop.stepdefs.then;

import io.cucumber.java.en.Then;
import org.springframework.beans.factory.annotation.Autowired;
import tw.waterballsa.bddworkshop.entity.Greeting;
import tw.waterballsa.bddworkshop.repository.GreetingRepository;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class Then_There_Is_N_Greeting_Records_In_The_Database {

    @Autowired
    private GreetingRepository greetingRepository;

    @Then("there is {int} greeting records in the database")
    public void validate(int expectedCount) {
        // Count all greeting records in the database
        long actualCount = 0;
        Iterable<Greeting> greetings = greetingRepository.findAll();
        
        // Count the number of greetings
        for (Greeting greeting : greetings) {
            actualCount++;
        }
        
        // Validate that the count matches the expected count
        assertEquals(expectedCount, actualCount, 
            "The number of greeting records in the database does not match the expected count");
    }
}

File: guessing-game-bdd-atomic/src/test/stepdefs/then/Then_There_Is_N_Greeting_Records_In_The_Database.ts
import { Then } from '@cucumber/cucumber';
import { INestApplication } from '@nestjs/common';
import { Repository } from 'typeorm';
import { Greeting } from '../../../main/greeting/entities/greeting.entity';
import { strict as assert } from 'assert';

export class Then_There_Is_N_Greeting_Records_In_The_Database {
    constructor(
        private readonly app: INestApplication
    ) {}

    @Then('there is {int} greeting records in the database')
    public async validate(expectedCount: number): Promise<void> {
        const greetingRepository = this.app.get<Repository<Greeting>>('GreetingRepository');

        // Count all greeting records in the database
        const actualCount = await greetingRepository.count();

        // Validate that the count matches the expected count
        assert.equal(actualCount, expectedCount,
            `The number of greeting records in the database does not match the expected count. Expected: ${expectedCount}, Actual: ${actualCount}`);
    }
}
